spec/services/hyrax/work_uploads_handler_spec.rb

```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+RSpec.describe Hyrax::WorkUploadsHandler do
+  subject(:service) { described_class.new(work: work) }
+
+  let(:uploads) { FactoryBot.create_list(:uploaded_file, 3) }
+  let(:work) { FactoryBot.valkyrie_create(:hyrax_work, :public) }
+
+  describe '#attach' do
+    it 'when no files are added returns uneventfully' do
+      expect { service.attach }
+        .not_to change { work.member_ids }
+        .from be_empty
+    end
+
+    context 'after adding files' do
+      before { service.add(files: uploads) }
+
+      it 'assigns FileSet ids synchronously' do
+        expect { service.attach }
+          .to change { work.member_ids }
+          .to contain_exactly(an_instance_of(Valkyrie::ID),
+                              an_instance_of(Valkyrie::ID),
+                              an_instance_of(Valkyrie::ID))
+      end
+
+      it 'creates persisted filesets' do
```

Comment:
It's a small ask, but could lines 27 and 32 be merged? I don't know if `:persisted?` is an attribute.  I think these two make sense as one singular expectation. (Not a blocker, but something to consider)

Reply:
i think i considered this fleetingly, wasn't sure if there was a pleasant way to get `have_attributes` to do this work, and decided i liked having separate tests/rspec messages for "did we save them" vs. "did we set the metadata". i also recall squashing visibility into L32, after initial writing it as yet another test case.

i guess our options for squashing are:

```ruby
have_attributes(title: ['image.jp2'], depositor: an_instance_of(String), visibility: 'open', "persisted?": true) # i'm not 100% sure this will work, but it seems plausible
# or
have_attributes(title: ['image.jp2'], depositor: an_instance_of(String), visibility: 'open', new_record: false)
```

if your vote is for one of the above over the current L27 test, i'll consider that decisive.

these tests _are_ a bit slow, as written, but we could add a `valkyrie_adapter: :test_adapter` flag and speed them up.

Reply:
i ended up making this use `:test_adapter` anyway. testing against generic Valkyrie is a benefit, so the speed factor is less of one now.